File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Models/Extensions.vb
Imports System.Runtime.CompilerServices

Public Module Extensions

    <Extension()>
    Public Function GetUniqueName(ByVal post As Post) As String
        If (post Is Nothing) Then
            Throw New ArgumentNullException("post")
        End If

        Dim ctx = HttpContext.Current
        Dim cacheKey = "post-staticurl-" + post.Id.ToString()

        Dim res As String = CStr(ctx.Cache(cacheKey))

        If (String.IsNullOrEmpty(res)) Then
            res = Regex.Replace(post.Title, "[^A-Za-z0-9\s]", String.Empty)
            res = Regex.Replace(res, "\s", "-")

            ctx.Cache.Insert(cacheKey, res, Nothing, DateTime.UtcNow.AddMinutes(20), Cache.NoSlidingExpiration)
        End If

        Return res
    End Function
End Module

File: aspnet-in-practice/08-09/CS/CoolMVBlog/Models/Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Caching;
using System.Text.RegularExpressions;

namespace CoolMvcBlog.Models
{
    public static class Extensions
    {
        public static string GetUniqueName(this Post post)
        {
            if (post == null)
                throw new ArgumentNullException("post");

            var ctx = HttpContext.Current;
            string cacheKey = "post-staticurl-" + post.Id;

            string res = ctx.Cache[cacheKey] as string;

            if (res == null)
            {
                res = Regex.Replace(post.Title, @"[^A-Za-z0-9\s]", string.Empty);
                res = Regex.Replace(res, @"\s", "-");

                ctx.Cache.Insert(cacheKey, res, null, DateTime.UtcNow.AddMinutes(20), Cache.NoSlidingExpiration);
            }

            return res;
        }

        public static void ForEach<T>(this IEnumerable<T> items, Action<T> action)
        {
            foreach (var item in items)
            {
                action(item);
            }
        }
    }
}